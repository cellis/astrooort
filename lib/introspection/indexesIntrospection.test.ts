import { Index, Indexes } from '@wmfs';
import { Client } from 'pg';
import connectTestDb from '../test/connectTestDb';
import { UserIndexes } from '../test/fixtures/MockData';
import {
  findIndexes,
  getTableFromFullyQualifiedPath,
  loadIndexes,
  normalizeTableNameForReserved,
  RawIndex,
} from './indexesIntrospection';

describe('indexes introspection', () => {
  let client: Client;
  let indexes: RawIndex[];
  beforeAll(async (done) => {
    client = await connectTestDb();
    indexes = await loadIndexes(client, 'superluminal');
    done();
  });

  afterAll(async () => client.end());

  describe('multiple schemas', () => {
    beforeEach(async (done) => {
      indexes = await loadIndexes(client, 'superluminal,superluminal_private');
      done();
    });
    it('finds indexes of all', () => {
      expect(indexes.length).toBeGreaterThan(0);
    });
  });
  describe('all indexes', () => {
    it('loads all indexes', () => {
      expect(indexes).toMatchSnapshot();
    });
  });

  describe('normalizeForPostgres', () => {
    it('returns \'user\' for "user"', () => {
      expect(normalizeTableNameForReserved('"user"')).toBe('user');
    });
  });

  describe('getTableFromFullyQualifiedPath', () => {
    it('returns \'user\' for superluminal."user"', () => {
      expect(getTableFromFullyQualifiedPath('superluminal."user"')).toBe(
        'user'
      );
    });
  });

  describe('find index', () => {
    it('finds the indexes for that model', () => {
      const foundIndexes: Indexes = findIndexes(
        indexes,
        'user',
        'superluminal'
      );

      const expected: Index = {
        columns: [['slug']],
        unique: true,
        method: 'btree',
      };
      expect(foundIndexes['user_pkey']).toEqual(expected);
    });
  });

  describe('user index', () => {
    it('normalizes the user index', async () => {
      expect(
        indexes.filter((index: any) => index.table_name === 'superluminal.user')
      ).toHaveLength(UserIndexes.length);
    });
  });
});
