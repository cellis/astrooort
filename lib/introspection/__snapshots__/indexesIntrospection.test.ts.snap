// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexes introspection all indexes loads all indexes 1`] = `
Array [
  Object {
    "index_keys": Array [
      "slug",
    ],
    "index_name": "user_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.user",
  },
  Object {
    "index_keys": Array [
      "full_text",
    ],
    "index_name": "user_full_text_idx",
    "is_primary": false,
    "is_unique": false,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.user",
  },
  Object {
    "index_keys": Array [
      "id",
    ],
    "index_name": "message_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.message",
  },
  Object {
    "index_keys": Array [
      "slug",
    ],
    "index_name": "transaction_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.transaction",
  },
  Object {
    "index_keys": Array [
      "user_id",
    ],
    "index_name": "transaction_user_id_idx",
    "is_primary": false,
    "is_unique": false,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.transaction",
  },
  Object {
    "index_keys": Array [
      "id",
    ],
    "index_name": "document_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.document",
  },
  Object {
    "index_keys": Array [
      "slug",
    ],
    "index_name": "shipment_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.shipment",
  },
  Object {
    "index_keys": Array [
      "\\"from\\"",
    ],
    "index_name": "shipment_from_user_slug_fk",
    "is_primary": false,
    "is_unique": false,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.shipment",
  },
  Object {
    "index_keys": Array [
      "\\"to\\"",
    ],
    "index_name": "shipment_to_user_slug_fk",
    "is_primary": false,
    "is_unique": false,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.shipment",
  },
  Object {
    "index_keys": Array [
      "id",
    ],
    "index_name": "product_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.product",
  },
  Object {
    "index_keys": Array [
      "slug",
    ],
    "index_name": "photo_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.photo",
  },
  Object {
    "index_keys": Array [
      "user_id",
    ],
    "index_name": "photo_user_id_idx",
    "is_primary": false,
    "is_unique": false,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.photo",
  },
  Object {
    "index_keys": Array [
      "slug",
    ],
    "index_name": "payment_details_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal_private",
    "table_name": "superluminal_private.payment_details",
  },
  Object {
    "index_keys": Array [
      "user_slug",
    ],
    "index_name": "account_pkey",
    "is_primary": true,
    "is_unique": true,
    "method": "btree",
    "schema_name": "superluminal",
    "table_name": "superluminal.account",
  },
]
`;
